babel/messages/extract.py

```diff
@@ -643,3 +650,43 @@ def extract_javascript(fileobj, keywords, comment_tags, options):
             funcname = token.value
 
         last_token = token
+
+
+def parse_template_string(template_string, fileobj, keywords, comment_tags, options):
+
+    prev_character = None
+    level = 0
+    inside_str = False
+    expression_contents = ''
+
+    for character in template_string[1:-1]:
+
+        if not inside_str and character in ('"', "'", '`'):
+            inside_str = character
+        elif inside_str == character and prev_character != r'\\':
+            inside_str = False
+
+        if level:
+            expression_contents += character
+
+        if not inside_str:
+
+            if character == '{' and prev_character == '$':
+                level += 1
+
+            elif level and character == '}':
+
+                level -= 1
+
+                if level == 0 and expression_contents:
+
+                    expression_contents = expression_contents[0:-1]
+
+                    fake_file_obj = io.BytesIO(expression_contents.encode())
+
+                    for item in extract_javascript(fake_file_obj, keywords, comment_tags, options):
```

Comment:
How does this interact with e.g. line numbering?

Reply:
line numbering has been fixed in #939. There is now also a test for that.